apollo-ios-pagination/Sources/ApolloPagination/AnyGraphQLPager.swift

```diff
@@ -54,35 +56,53 @@ public class AnyGraphQLQueryPager<Model> {
   >(
     pager: Pager,
     initialTransform: @escaping (InitialQuery.Data) throws -> Model,
-    nextPageTransform: @escaping (NextQuery.Data) throws -> Model
+    pageTransform: @escaping (NextQuery.Data) throws -> Model
   ) where Model: RangeReplaceableCollection, Model.Element == Element {
     self.init(
       pager: pager,
-      transform: { initialData, nextData in
+      transform: { previousData, initialData, nextData in
+        let previous = try previousData.flatMap { try pageTransform($0) }
         let initial = try initialTransform(initialData)
-        let next = try nextData.flatMap { try nextPageTransform($0) }
-        return initial + next
+        let next = try nextData.flatMap { try pageTransform($0) }
+        return previous + initial + next
       }
     )
   }
 
-  /// Subscribe to new pagination `Output`s.
-  /// - Parameter completion: Receives a new `Output` for the consumer of the API.
-  /// - Returns: A `Combine` `AnyCancellable`, such that the caller can manage its own susbcription.
-  @discardableResult public func subscribe(completion: @escaping (Output) -> Void) -> AnyCancellable {
+  deinit {
+    pager.cancel()
+  }
+
+  @discardableResult public func subscribe(completion: @MainActor @escaping (Output) -> Void) -> AnyCancellable {
     guard let _subject else { return AnyCancellable({ }) }
```

Comment:
Does this `AnyCancellable` need to be stored in `cancellables`? Even though it doesn't do anything, it should still get a cancellation when the pager is cancelled, shouldn't it?

Reply:
I think this is a situation where we're going to have to decide what pattern we want to follow. We can either return the cancellable, or store it internally, but we probably shouldn't do both.

Reply:
It's a tough one; I'm inclined to agree with you on this. Perhaps we can mark this as an area for future consideration; thinking outside of the FRP-world, I generally try to contain all of those concepts so that folks that aren't using an FRP paradigm don't suddenly have to juggle these concepts.

🤔

Reply:
>  I generally try to contain all of those concepts so that folks that aren't using an FRP paradigm don't suddenly have to juggle these concepts.

Yeah, I agree with that in general. But I'm thinking ahead to Apollo 2.0, and adopting modern swift concurrency is one of our biggest goals (async/await support is a highly requested features). But because of the multiple completion handler calls (from the cache, then server) we make, that is likely going to require us to expose `Combine` streams and other FRP concepts as return values for network requests.

As a singular feature library, I think that philosophy is usually correct, but Apollo is more of an entire SDK. Between model generation, caching, and networking request handling, it is almost always integrated into the heart of your project's infrastructure. Apple is steering the Swift language and `SwiftUI` development towards FRP with `Combine` and `async/await`. So I think it's appropriate for an SDK like Apollo to follow similar API design concepts as them.

---

This pagination module is an interesting case though, because it's being released prior to the 2.0 redesign. So it would be the first place we are exposing those concepts. On one hand, that gives us an opportunity to test the waters with this, and see how users respond; on the other, it may be inconsistent with the API design language/style of Apollo iOS.

So now, I'm leaning back towards exposing the `Cancellable`! 🙄 

CC: @bignimbus @apollographql/prometheus in case any of you have an opinion on this API design philosophy question.

Reply:
@AnthonyMDev perhaps as a compromise:

```swift
  public func subscribe(completion: @MainActor @escaping (Output) -> Void) -> AnyCancellable {
    guard let _subject else { return AnyCancellable({ }) }
    return _subject.compactMap({ $0 }).sink { result in
      Task {
        await completion(result)
      }
    }
  }

  public func subscribe(completion: @MainActor @escaping (Output) -> Void) {
    subscribe(completion: completion).store(in: &cancellables)
  }
```